### DISCUSSION TITLE
Inaccurate contact between rod and sphere

### DISCUSSION THREAD

**Discussion by carlo-alessi** [GazzolaLab/PyElastica#264] [Ideas]
Dear Elastica developers,

I noticed a minor problem in PyElastica during the simulation of the contact between a Rod and a Sphere with

```
simulator.connect(rod, sphere).using(ExternalContact, k=k, nu=nu,
				     velocity_damping_coefficient=1e5,
				     friction_coefficient=0.5).
```

In particular, the simulation is inaccurate when the contact happens with the cross-section of the last node of the rod facing the sphere (say frontal contact). This could be because the function `_calculate_contact_forces_rod_rigid_body()` computes the contact forces as a function of the rod radius and sphere radius. The implementation is effective when contact is "lateral" but fails in the case of frontal contact. The problem is only noticeable for large k (e.g., k=1000).

Could you please fix this minor problem or suggest some temporary workaround? :)

Best regards,
Carlo Alessi

**Comment by Ali-7800** [comment]
Hi @noobcode,

Currently PyElastica only supports rod-rod and rod-cylinder external contact. You are right about the inaccuracy in the frontal contact between the rod and the cylinder, this is because we treat the rod elements as tubes instead of cylinders as they should be.  We're currently reworking and expanding the contact module to fix this issue and to include arbitrary rigid bodies such as spheres.
